src/goose_plugins/toolkits/complexity_analyzer.py

```diff
@@ -0,0 +1,174 @@
+
+        Args:
+            code (str): The Python code as a string to analyze.
+
+        Returns:
+            int: The Cyclomatic Complexity of the code.
+        """
+        try:
+            complexity_list = rc.cc_visit(ast.parse(code))
+            total_complexity = 0
+
+            # Iterate over each item in the complexity list
+            for item in complexity_list:
+                if hasattr(item, "complexity"):
+                    # Add complexity of the function or class's top-level complexity
+                    total_complexity += item.complexity
+
+                # For classes, add complexity of methods if any
+                if hasattr(item, "methods"):
+                    for method in item.methods:
+                        total_complexity += method.complexity
+            return total_complexity
+        except Exception as e:
+            print(e)
+            self.notifier.log(f"Error calculating cyclomatic complexity: {str(e)}")
+            return 0
+
+    @tool
+    def halstead_complexity(self, code: str) -> dict:
+        """Calculate Halstead Complexity metrics of the given Python code.
+
+        Args:
+            code (str): The Python code as a string to analyze.
+
+        Returns:
+            dict: A dictionary containing the Halstead metrics, including 'halstead_volume'.
+        """
+        from radon.metrics import h_visit
+
+        try:
+            halstead_report = h_visit(code)
+            return {
+                "halstead_volume": halstead_report.total.volume,
+                "details": {
+                    "vocabulary": halstead_report.total.vocabulary,
+                    "length": halstead_report.total.length,
+                    "calculated_length": halstead_report.total.calculated_length,
+                    "difficulty": halstead_report.total.difficulty,
+                    "effort": halstead_report.total.effort,
+                    "time": halstead_report.total.time,
+                    "bugs": halstead_report.total.bugs,
+                },
+            }
+        except Exception as e:
+            print(e)
+            self.notifier.log(f"Error calculating Halstead complexity: {str(e)}")
+            return {}
+
+    @tool
+    def maintainability_index(self, code: str) -> int:
+        """Calculate the Maintainability Index of the given Python code.
+
+        Args:
+            code (str): The Python code as a string to analyze.
+
+        Returns:
+            int: The Maintainability Index of the code.
+        """
+
+        try:
+            mi_score = rm.mi_visit(code, multi=True)
+            return mi_score
+        except Exception as e:
+            print(e)
+            self.notifier.log(f"Error calculating maintainability index: {str(e)}")
+            return 0
+
+
+from unittest.mock import MagicMock
+
+toolkit = CodeComplexityToolkit(notifier=MagicMock())
+result = toolkit.analyze_complexity("/Users/jevan/personal/algos")
```

Comment:
will this break when someone else runs this? as they dont have the same file path as you?

Reply:
yea i was just tryna test the code, lemme remove it

Reply:
sorted now @blackgirlbytes